### PR TITLE
Normalize quaternions from msgs before Ogre use

### PR DESCRIPTION
**Pull request by rhaschke**
This is a follow-up to #1167, which added tests for invalid quaternions that caused crashes in Ogre.
However, instead of rejecting those invalid quaternions, I propose to normalize them (or use identity if a zero quaternion is passed). Actually this was already the default behaviour of `FrameManager::transform()` which is used by most routines anyway.
However, in rare cases ROS quaternions were used directly to set an `Ogre::Quaternion`, which then indeed requires normalization and special handling of the zero quaternion.

This PR reverts most changes introduced in #1167 and thus fixes the regression not handling non-normalized quaternions gracefully: #1178, https://github.com/ros-planning/moveit/issues/732.

### CODE REVIEW

**Comment by davetcoleman** [comment]
I just ran into this issue today, wasted ~4 hours trying to understand why some machines it was working until I upgraded them all.

File: src/rviz/default_plugin/covariance_visual.cpp [hunk 1]
```diff
@@ -299,7 +300,10 @@ void CovarianceVisual::setCovariance( const geometry_msgs::PoseWithCovariance& p
   updateOrientationVisibility();
 
   // store orientation in Ogre structure
-  Ogre::Quaternion ori(pose.pose.orientation.w, pose.pose.orientation.x, pose.pose.orientation.y, pose.pose.orientation.z);
+  Ogre::Quaternion ori;
+  if (!normalizeQuaternion(pose.pose.orientation, ori))
+    ROS_WARN("invalid quaternion (zero length)");
```
Comment (davetcoleman) [inline src/rviz/default_plugin/covariance_visual.cpp]: nit: capitalize Invalid here and in rest of this PR

File: src/rviz/validate_quaternions.h [hunk 2]
```diff
@@ -39,65 +39,104 @@
...
-  return validateQuaternions( msg.w, msg.x, msg.y, msg.z );
+  return std::abs(quaternionNorm2(w, x, y, z) - 1.0) < 10e-3;
 }
 
-inline bool validateQuaternions( const geometry_msgs::Pose &msg )
+inline bool normalizeQuaternion(double& w, double& x, double& y, double& z)
 {
-  return validateQuaternions( msg.orientation );
+  float norm2 = quaternionNorm2(w, x, y, z);
+  if (norm2 < 10e-3f)
```
Comment (davetcoleman) [inline src/rviz/validate_quaternions.h]: can you add a comment here why this threshold was chosen?

File: src/rviz/validate_quaternions.h [hunk 3]
```diff
@@ -39,65 +39,104 @@
...
-  for ( ; it != end; ++it )
+  return validateQuaternion(msg.orientation);
+}
+
+inline bool validateQuaternion(const geometry_msgs::PoseStamped &msg)
+{
+  return validateQuaternion(msg.pose.orientation);
+}
+
+
```
Comment (davetcoleman) [inline src/rviz/validate_quaternions.h]: one line break

**Comment by davetcoleman** [comment]
I've tested this patch locally and it fixes the issue for me, thanks again

**Comment by tfoote** [comment]
I'm not sure that I agree with the logic to just "normalize" the quaternion on input. Sending an unnormalized quaternion is like sending a NaN for a float value. Just pushing it to the nearest valid float real value is not representing the same thing. 

When doing quaternion math floating point arithmetic errors can induce quaternions that are non-normalized and they do periodically need to be renormalized. Renormalizing at every step is a very high overhead, normalizing never will end up with NaN values. There is a balance in between to two extremes. This is typically done in a specific time in processing. Often I would expect it to be done shortly before being output from an algorithm not on the input. As the input will not have semantic knowledge of the previous potential cause of non-normalization. And I think it's much more correct for input validation instead of input normalization.

Related to this arbitrarily normalizing all inputs can cause completely invalid quaternions (such as unset) to become valid and be rendered. Normalization can only be considered valid if the original is close to normalized, or in a known state of non-normalization. (for example unitialized with all zeros) I can certainly take 4 numbers and turn them into a normalized quaternion, but if they still represent the same underlying information is not clear. Thus it would be better to reject the invalid data rather than pushing it into the right form.

Another example of this might be if the throttle commands for a vehicle are on the range 0 to 1. If I receive a command of 1.1 what does that mean? It's undefined as it's not in the interval, it could be someone trying to dial it up to 11. Or it could be someone sending 1.1 out of a range on 100. The meaning of a non-normalized quaternions is similarly undefined. I would strongly advocate for resolving this issue by making sure that all published quaternions are valid rather than accepting invalid quaternions by auto-renormalizing them.

**Comment by rhaschke** [comment]
@tfoote To some extend I can agree to your argumentation. However, it's a plain fact, that rviz did (and still does if allowed to) normalize quaternions incoming via ROS msgs and now it simply rejects them. This dramatically changes behaviour of rviz and - besides others - breaks the plain interactive marker tutorial (which could be thought of a best-practice example).
So, I'm fine to change the semantics in some future. But, please give us some time for transition and don't break a released tool.
Also, we should provide some tools to normalize quaternions in a `ROS pose msg`. Then it becomes feasible to normalize at the output generator (instead of the input in rviz).

**Comment by tfoote** [comment]
We're looking at restoring the permissive behavior to restore existing behavior, but adding warnings. Anywhere that is known to be sending unnormalized quaternions should be ticketed for resolving at the source. Interactive markers tutorials or libraries should not be sending unnormalized quaternions. Please open tickets for anything that's having trouble right now while we're working to resolve this.

**Comment by wjwwood** [comment]
> However, it's a plain fact, that rviz did (and still does if allowed to) normalize quaternions incoming via ROS msgs

Is that true? Looking at #1167, I don't see where code that did normalize incoming marker message quaternions was removed. Perhaps you mean in other places?

> This dramatically changes behaviour of rviz and - besides others - breaks the plain interactive marker tutorial (which could be thought of a best-practice example).
So, I'm fine to change the semantics in some future. But, please give us some time for transition and don't break a released tool.

I can't speak for @tfoote, but we're on the same page here, the disruptive change in behavior was unexpected and we're going to revert it and do something about it. I think the idea of requiring valid quaternions would be a future change or at least would be opt-in if back ported.

I think it's worth pursuing to get valid quaternions in rviz (either fixed on the rviz side or on the publishing side), because there have been other issues which attribute crashes to invalid quaternions, but we haven't conclusively linked those two things yet, e.g.:

https://github.com/ros-visualization/rviz/issues/1082#issuecomment-354312484

**Comment by dhood** [comment]
I don't believe we ever normalised un-normalised quaternions before 1167, but @rhaschke pointed out that, previously, null quaternions would get set to identity in `FrameManager::transform()` while now they [are rejected before they get to that point](https://github.com/ros-visualization/rviz/pull/1139/files#diff-bc9015d4b6ff8d1c30e5f237d306fcb3R297), which may be what they were referring to wrt "normalising" in the past

**Comment by gavanderhoorn** [comment]
The unexpected breakage of RViz is not nice, but patching up invalid data just so it can be rendered makes no sense to me, so my vote would be for the warn/error-and-discard option.

That would be similar to TF and other incoming msgs that can't be transformed properly: they're discarded and a warning/error with sufficient visibility is output (ie: "Dropped N% of msgs so far, ..").

Normalising unconditionally will probably just mean that other nodes will crash/error out on the same data, which could be equally confusing ("but RViz does accept my quaternion?").

**Comment by rhaschke** [comment]
To clarify my point: `rviz` currently adapts invalid quaternions in at least two central places (since ages):
* null quaternions are replaced by identity quaternions in [`rviz::FrameManager::transform`](https://github.com/ros-visualization/rviz/blob/1.12.14/src/rviz/frame_manager.cpp#L243)
* quaternions are normalized in [`tf::Matrix3x3::setRotation`](https://github.com/ros/geometry/blob/indigo-devel/tf/include/tf/LinearMath/Matrix3x3.h#L150) ([same in tf2](https://github.com/ros/geometry2/blob/indigo-devel/tf2/include/tf2/LinearMath/Matrix3x3.h#L150))
called from [`rviz::FrameManager::transform`](https://github.com/ros-visualization/rviz/blob/1.12.14/src/rviz/frame_manager.cpp#L246)

So, its not only `rviz`, which re-normalized quaternions, but that's already part of more basic `tf library`!

I fully understand the wish to handle normalization once at generation time, but current practice is fundamentally different. I guess, if you enable a warning in [`tf`](https://github.com/ros/geometry2/blob/indigo-devel/tf2/include/tf2/LinearMath/Matrix3x3.h#L150) you will get hundreds of warnings. It's common practice to create quaternions like `(1,0,0,1)` to represent 90° rotations about some axes and then relying on normalization.

**Comment by rhaschke** [comment]
I addressed @davetcoleman's comments.

**Comment by davetcoleman** [comment]
I agree we should require all input to be normalized *in the future* but for now it should just produce warnings instead of breaking Rviz worldwide. I have found regressions in several different uses of Rviz, including MoveIt!.

**Comment by dhood** [comment]
@rhaschke thank you for clarifying that tf will normalise the quaternions inside `FrameManager::transform`. In light of this it makes sense to continue normalising in the missing places as you have proposed.

Since we still want warnings for unnormalised quaternions I'll remove the reverting of 1167 from this PR in favour of https://github.com/ros-visualization/rviz/pull/1182, and this PR will just focus on normalising the quaternions from ROS msgs that are not otherwise normalised.

Additionally I'll invert the logic to consider null quaternions valid (users being lazy) and unnormalised ones invalid (potentially sign of an error)

**Comment by dhood** [comment]
@rhaschke thank you for kick starting this and for giving permission for this branch to be adapted by us.

As I mentioned, this PR now addresses the remaining places in the codebase @rhaschke identified where quaternions may be null/unnormalised. It addresses the issue that initially prompted rejecting invalid quaternions (https://github.com/ros-visualization/rviz/issues/1137) and potentially others. It therefore removes the need for us to reject invalid quaternions (we now give warnings instead: #1182)

I've minimised the diff to simplify review, including removing printing of warnings for invalid quaternions that are covered by #1182. `normalizeQuaternions` still sets null quaternions to identity and normalises all others, but no longer returns a boolean; callers must use `validateQuaternions` to determine if a warning should be printed (and to get the invalid quaternion's magnitude logged to debug).

@rhaschke The previous state of this PR had [map_display rejecting null quaternions](https://github.com/ros-visualization/rviz/pull/1179/commits/3dc2e74dec48123c716e5262c79a6999bf11ab7c#diff-218addc62ac0b6e16ffb70cffef3bfe8R663) which is no longer the case since I imagine it was unintentional, please correct me if that's not the case

**Comment by dhood** [comment]
@ros-pull-request-builder retest this please

**Comment by dhood** [comment]
@ros-pull-request-builder retest this please

(issues with the git refs)

**Comment by wjwwood** [comment]
Obviously need to get the CI passing or verify it locally, otherwise lgtm.

File: src/rviz/validate_quaternions.h [hunk 4]
```diff
@@ -40,104 +40,146 @@
...
+inline float normalizeQuaternion( float& w, float& x, float& y, float& z )
 {
-  float norm2 = quaternionNorm2(w, x, y, z);
-  if (norm2 < 10e-3f)
-    return false;
-  norm2 = std::sqrt(norm2);
+  if ( 0.0f == x && 0.0f == y && 0.0f == z && 0.0f == w )
+  {
+    w = 1.0f;
+    x = y = z = 0.0f;
```
Comment (rhaschke) [inline src/rviz/validate_quaternions.h]: This line is superfluous. x,y,z are zero already.

File: src/rviz/validate_quaternions.h [hunk 5]
```diff
@@ -40,104 +40,146 @@
...
+inline double normalizeQuaternion( double& w, double& x, double& y, double& z )
 {
-  float norm2 = quaternionNorm2(w, x, y, z);
-  if (norm2 < 10e-3f)
-    return false;
-  norm2 = std::sqrt(norm2);
+  if ( 0.0 == x && 0.0 == y && 0.0 == z && 0.0 == w )
+  {
+    w = 1.0;
+    x = y = z = 0.0;
```
Comment (rhaschke) [inline src/rviz/validate_quaternions.h]: Superfluous again.

File: src/rviz/validate_quaternions.h [hunk 6]
```diff
@@ -40,104 +40,146 @@
...
   q.y = msg.y;
   q.z = msg.z;
-  return normalizeQuaternion(q);
+  return normalizeQuaternion( q );
 }
 
-
 template<typename T>
-inline bool validateQuaternions(const T &vec)
+inline bool validateQuaternions( const std::vector<T> &vec )
```
Comment (rhaschke) [inline src/rviz/validate_quaternions.h]: I more like the old version as it is more compact and easier to grasp ;-)

**Comment by rhaschke** [comment]
> @rhaschke The previous state of this PR had map_display rejecting null quaternions which is no longer the case since I imagine it was unintentional, please correct me if that's not the case.

@dhood I think, converting null quaternions in a message to `identity` is what the user expects.

File: src/rviz/validate_quaternions.h (hunk 4 above)
Comment (dhood) [reply src/rviz/validate_quaternions.h]: gah, thanks for pointing that out: df6dfd4

File: src/rviz/validate_quaternions.h (hunk 6 above)
Comment (dhood) [reply src/rviz/validate_quaternions.h]: yeah, I know, but I'm trying to keep unnecessary changes out for this PR 😃

**Comment by dhood** [comment]
> The current proposal explicitly separates validation and normalization which is fine if the additional overhead is accepted.

Yeah, for most of these msgs they're not being normalised elsewhere so it's roughly the same overhead as the validateQuaternions + tf normalisation in other places

The devel jobs are behaving now so we'll just wait for that to turn over. Appreciate your contributions/insight on this @rhaschke

**Comment by rhaschke** [comment]
> Thank you for giving permission for this branch to be adapted by us.

@dhood How did you managed to push to my PR branch? I'm trying to do the same in another project.

**Comment by VictorLamoine** [comment]
I think we should reject quaternions that are not valid in the future.
I understand it will break a lot of code but this clearly needs to be fixed at the source; quaternions must be defined.

I'm in favour of:
- Normalizing/initializing incoming quaternions AND printing warnings about invalid quaternions
- Reporting and fixing of as much packages as possible during a defined time (6 months? 1 year?)
- Rejecting invalid quaternions, hoping that most packages are already fixed.

It will take some time to fix the code base but the effort is worth it.
What is the plan?

**Comment by VictorLamoine** [comment]
Did someone test the overhead of this patch? eg: when displaying a large number of markers?

**Comment by dhood** [comment]
OK, looks like we're all in agreement.

Summarising the thread: invalid quaternions will be permitted for now to not break displays; publishers of invalid quaternions should be reported/updated; invalid quaternions will be rejected in the future.

> What is the plan?

Pretty much as you said, @VictorLamoine. This is the timeline to get to the end goal of not normalising inputs:

- Now: normalizing/initializing incoming quaternions AND printing warnings about invalid quaternions, with the 'once' log filter to ease into it.
- After 6 months: more + stronger warnings to encourage the remaining offenders.
- After 1 year: reject invalid quaternions with a warning.

Regarding the overhead, we can consider adding an option to displays to assume valid quaternions if there are use cases for which the overhead needs to be avoided.

**Comment by rhaschke** [comment]
@dhood @VictorLamoine In order to enforce valid quaternions on the generating side, you should provide tools to:

- [ ]  correctly initialize an identity quaternion in [corresponding ROS msgs](http://docs.ros.org/api/geometry_msgs/html/msg/Quaternion.html) by default. Currently, the default is `(0,0,0,0)`.
- [ ] centrally provide a tool function to normalize a [`quaternion msg`](http://docs.ros.org/api/geometry_msgs/html/msg/Quaternion.html).

**Comment by wjwwood** [comment]
> correctly initialize an identity quaternion in corresponding ROS msgs by default. Currently, the default is (0,0,0,0).


That's not possible because there are no default values for generated messages in ROS atm. In ROS 2 we have this (specifically for this case with quaternions), but I don't think there are plans to back port it, but with some community interest and help it might happen.

**Comment by rhaschke** [comment]
> > correctly initialize an identity quaternion in corresponding ROS msgs by default. Currently, the default is (0,0,0,0).

> That's not possible because there are no default values for generated messages in ROS atm.

Actually, I was afraid of this. For me, that's a clear reason that generation-side normalization is currently not feasible in a comfortable fashion.

**Comment by wjwwood** [comment]
That's always been the case though. I don't see why requiring people to initialize their quaternions is a problem. Also for the special case of all zeros, we can continue to change that to identity for people, since that's not as expensive as always normalizing. That case, in my mind, is separate from giving unnormalized quaternions.

File: src/rviz/validate_quaternions.h [hunk 7]
```diff
@@ -71,10 +71,11 @@ inline float normalizeQuaternion( float& w, float& x, float& y, float& z )
   }
   float norm2 = quaternionNorm2( w, x, y, z );
   norm2 = std::sqrt( norm2 );
-  w /= norm2;
-  x /= norm2;
-  y /= norm2;
-  z /= norm2;
+  float invnorm = 1.0f / norm2;
```
Comment (rhaschke) [inline src/rviz/validate_quaternions.h]: What's the benefit of this? [Multiplication and division have identical costs on the CPU](https://stackoverflow.com/questions/226465/should-i-use-multiplication-or-division), haven't they?

**Comment by rhaschke** [comment]
> Also for the special case of all zeros, we can continue to change that to identity for people.

Agreed.

**Pull request merged**